src/test/java/io/jenkins/plugins/audit/listeners/UserCreationListenerTest.java

```diff
@@ -0,0 +1,161 @@
+
+        // credentials of four Jenkins accounts
+        USERS.put("alice", "alicePassword");
+        USERS.put("bob", "bobPassword");
+        USERS.put("charlie", "charliePassword");
+        USERS.put("debbie", "debbiePassword");
+
+        client = j.createWebClient();
+        logout(client);
+
+        app = ListAppender.getListAppender("AuditList").clear();
+    }
+
+    @After
+    public void teardown() {
+        app.clear();
+    }
+
+    @Issue("JENKINS-54088")
+    @Test
+    @Parameters({
+            "1, alice, alicePassword",
+            "1, bob, bobPassword",
+            "1, charlie, charliePassword",
+            "1, debbie, debbiePassword"
+    })
+    public void testUserCreationFromRealm(int expectedCount, String username, String password) throws Exception {
+        List<LogEvent> events = app.getEvents();
+        assertEventCount(events, 0);
+
+        HudsonPrivateSecurityRealm realm = new HudsonPrivateSecurityRealm(false, false, null);
+        j.jenkins.setSecurityRealm(realm);
+
+        User user = realm.createAccount(username, password);
+        user.save();
+
+        assertEventCount(events, expectedCount);
+    }
+
+    @Issue("JENKINS-54088")
+    @Test
+    public void testUserCreationFromSignUp() throws Exception {
+        List<LogEvent> events = app.getEvents();
+        assertEventCount(events, 0);
+
+        HudsonPrivateSecurityRealm realm = new HudsonPrivateSecurityRealm(true, false, null);
+        j.jenkins.setSecurityRealm(realm);
+
+        SignupPage signup = new SignupPage(client.goTo("signup"));
+        signup.enterUsername("debbie");
+        signup.enterPassword(USERS.get("debbie"));
+        signup.enterFullName("Debbie User");
+        HtmlPage success = signup.submit(j);
+
+        // user creation via a jenkins signup also automatically logs the user in
+        assertEventCount(events, 2);
+
+        // verify a login event occurred
+        client.executeOnServer(() -> {
+            Authentication a = Jenkins.getAuthentication();
+            assertEquals("debbie", a.getName());
+
+            return null;
+        });
+
+        assertThat(success.getElementById("main-panel").getTextContent(), containsString("Success"));
+        assertEquals("Debbie User", realm.getUser("debbie").getDisplayName());
+    }
+
+    @Issue("JENKINS-54088")
+    @Test
+    public void testUserCreationAndLoginFromRealm() throws Exception {
+        List<LogEvent> events = app.getEvents();
+        assertEventCount(events, 0);
+
+        HudsonPrivateSecurityRealm realm = new HudsonPrivateSecurityRealm(false, false, null);
+        j.jenkins.setSecurityRealm(realm);
+
+        User u1 = realm.createAccount("charlie", USERS.get("charlie"));
+        u1.save();
+        client.login("charlie", USERS.get("charlie"));
+
+        // verify the audit event log messages as user creation and user login events
+        StructuredDataMessage logMessageOne = (StructuredDataMessage) events.get(0).getMessage();
```

Comment:
You can use the [get method](https://logging.apache.org/log4j/2.x/log4j-api/apidocs/org/apache/logging/log4j/message/MapMessage.html#get(java.lang.String)) to look up keys directly to verify their values. Or did you already try that?

Reply:
the current messages are of the form [createUser userId="charlie"]  or [login userId="debbie"] , while I can use the `get()` method to get the `userId` since its a key-value pair, I can't use the method to return 'createUser' or 'login' directly, which is why I was using the `contains()` method

Reply:
I could use the `get()` method in the `UserLoginListenerTest` and `UserLogoutListenerTest` class though, since there are some tests that evaluate direct comparison with the `userId` field. will refactor those classes in abit.

Reply:
Oh, is get() only for request context attributes?

Reply:
no i only meant if more details/data was available in the messages (besides what's in the current format) I'd be able to use the `get()` method more since there'd be alot of key-value pairs to work with, as oppose to there just being only one right now (i.e. `userId`)

Reply:
Oh, so then you'd make the log assertions more detailed when there's more keys to assert?

Reply:
yup